File: chroma-league/src/main/java/net/booone/chromaleague/hud/parts/ProgressBar.java
package net.booone.chromaleague.hud.parts;

import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.Color;
import net.booone.razersdk.sdk.RzKey;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ProgressBar extends SimpleFrame {

    public ProgressBar(final List<RzKey> progressBar, final Integer percent, final Color color) {
        super(getBarPercent(progressBar, percent), color);
    }

    public ProgressBar(final LinkedHashMap<RzKey, Color> progressBar, final Integer percent) {
        super(progressBar.entrySet().stream()
                .limit(indexToFill(progressBar.size(), percent))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
    }

    private static List<RzKey> getBarPercent(final List<RzKey> progressBar, final Integer percent) {
        return progressBar.subList(0, indexToFill(progressBar.size(), percent));
    }

    public static int indexToFill(int progressBarSize, Integer percent) {
        if (percent >= 100) {
            return progressBarSize;
        }
        return Double.valueOf(progressBarSize * (percent * 0.01)).intValue();
    }
}

File: src/hud/parts/ProgressBar.ts
import {Color} from '../../razer-sdk/color/Color.js';
import {SimpleFrame} from '../../razer-sdk/animation/SimpleFrame.js';
import {RzKey} from '../../razer-sdk/sdk/RzKey.js';

export class ProgressBar extends SimpleFrame {
    constructor(progressBar: readonly RzKey[], percent: number, color: Color);
    constructor(progressBar: Map<RzKey, Color>, percent: number);
    constructor(
        progressBar: readonly RzKey[] | Map<RzKey, Color>,
        percent: number,
        color?: Color,
    ) {
        if (progressBar instanceof Map) {
            const limit = indexToFill(progressBar.size, percent);
            const limited = new Map<RzKey, Color>();
            let i = 0;
            for (const [key, val] of progressBar) {
                if (i >= limit) break;
                limited.set(key, val);
                i++;
            }
            super(limited);
        } else {
            super(getBarPercent(progressBar, percent), color!);
        }
    }
}

function getBarPercent(progressBar: readonly RzKey[], percent: number): readonly RzKey[] {
    return progressBar.slice(0, indexToFill(progressBar.length, percent));
}

export function indexToFill(progressBarSize: number, percent: number): number {
    if (percent >= 100) {
        return progressBarSize;
    }
    return Math.trunc(progressBarSize * (percent * 0.01));
}
